### PR TITLE
[MM-35222] Fix DoGuestRolesCreationMigration

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

This migration was failing during scheme validation since we were setting both channel and team roles.
PR fixes this by only setting the role based on the scope for the scheme.

I am not very experienced with our permissions system so please let me know if you think this should be addressed differently.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-35222

```release-note
NONE
```

### CODE REVIEW

**Review by hahmadia — APPROVED** [review]
Tagging Scott Bishel as well just to make sure I've not missed anything. Makes sense to me otherwise.

**Comment by amyblais** [comment]
@streamer45 Is this needed for v5.35?

**Comment by streamer45** [comment]
> @streamer45 Is this needed for v5.35?

@amyblais 
I think it would be good to include it. It's not a regression as far as I can see but we can probably prevent some issues by having this fixed sooner rather than later.

File: app/migrations.go [hunk 1]
```diff
@@ -191,34 +191,36 @@ func (a *App) DoGuestRolesCreationMigration() {
...
+					SchemeManaged: true,
+				}
+
+				if savedRole, err := a.Srv().Store.Role().Save(teamGuestRole); err != nil {
+					mlog.Critical("Failed to create new guest role for custom scheme.", mlog.Err(err))
+					allSucceeded = false
+				} else {
+					scheme.DefaultTeamGuestRole = savedRole.Name
+				}
+			} else if scheme.Scope == model.SCHEME_SCOPE_CHANNEL {
```
Comment (jespino) [inline app/migrations.go:209]: I think this `else if` is wrong. If I remember correctly (and If I'm interpreting correctly the code in `store/sqlstore/scheme_store.go:115`), the `DefaultChannelGuestRole` is going to be part any `SCHEME_SCOPE_TEAM` and `SCHEME_SCOPE_CHANNEL` so... the `DefaultTeamGuestRole` is only for the `SCHEME_SCOPE_TEAM` but the `DefaultChannelGuestRole` is for both.

Comment (streamer45) [reply app/migrations.go:209]: That makes sense. That means we only need the check on the team scope.

Comment (sbishel) [reply app/migrations.go:209]: Yes, `SCHEME_SCOPE_TEAM` should get the Channel roles as well. So the `else if` isn't needed.

File: app/migrations.go [hunk 2]
```diff
@@ -220,12 +222,12 @@ func (a *App) DoGuestRolesCreationMigration() {
...
 				scheme.DefaultChannelGuestRole = savedRole.Name
 			}
+		}
 
-			_, err := a.Srv().Store.Scheme().Save(scheme)
-			if err != nil {
-				mlog.Critical("Failed to update custom scheme.", mlog.Err(err))
-				allSucceeded = false
-			}
+		_, err := a.Srv().Store.Scheme().Save(scheme)
```
Comment (jespino) [inline app/migrations.go:227]: Now you are saving this for every scheme not only for the ones that needs the migration. It should go inside the if block checking for empty DefaultChannelGuestRole or DefaultTeamGuestRole.

Comment (streamer45) [reply app/migrations.go:227]: Oh dear, I've messed up with the brackets. Thanks.

**Review by jespino — APPROVED** [review]
LGTM

File: app/migrations.go (hunk 2 above)
Comment (jespino) [reply app/migrations.go]: yes, gofmt is great until it fixes something that you didn't expect to be fixed :)

**Comment by lindy65** [comment]
@streamer45 - I've asked in QA channel for an SDET to review this PR and test once ticket is resolved as it involves restarting the server and checking `systems` table.

**Review by sbishel — APPROVED** [review]

**Pull request merged**